// webpack-php-mock/example/app/components/shopareas/list/List.jsx
import React from 'react';
import './List.less';

var template = require('./list.tpl');

class AreaList extends React.Component {
    constructor(props) {
        super(props);
        this.state = {data: []};
    }

    render() {
        return (
            <div className="area-list panel panel-default" onClick={this.props.onClick}>
                <div className="list-table" 
                    dangerouslySetInnerHTML={
                        {__html: template.fetch({
                            detail: this.props.detail,
                            data: this.state.data, 
                            mode: this.state.mode,
                            aoiid: this.state.aoiid
                        })}
                    }
                >
                </div>
                <div className="detail-btn btn btn-default">
                    <i data-node="detail" className="to-detail glyphicon glyphicon-menu-right"></i>
                </div>
            </div>
        );
    }

    setData(data) {
        this.setState({data: data.data});
        this.setState({mode: data.mode});
        this.setState({aoiid: data.aoiid});
    }
}

export default AreaList;
