// webpack-php-mock/example/app/components/shopcost/list/list.jsx
import React from 'react';
import ButtonNav from 'components/common/buttonnav/ButtonNav';
import './list.less';

var template = require('./list.tpl');
var poptpl = require('./pop.tpl');

class CastList extends React.Component {
    constructor(props) {
        super(props);
        this.state = {data: []};
    }

    render() {
        return (
            <div data-node="cost-list" ref="el" className="cost-list panel panel-default" onClick={this.props.onClick}>
                <div className="list-table" 
                    dangerouslySetInnerHTML={
                        {__html: template.fetch({detail: this.props.detail, data: this.state.data, shopId: this.state.shopId})}
                    }
                >
                </div>
            </div>
        );
    }
    componentWillReceiveProps(nextProps) {
        var castData = nextProps.castData;

        if (castData) {
            var shopId = castData.shopId;
            this.setState({shopId: shopId});
        }
        else {
            this.setState({shopId: -1});
        }
    }

    componentDidMount() {
        this.$el = $(this.refs.el.getDOMNode());
    }

    componentDidUpdate() {
        var me = this;
        me.$el.find('[data-toggle="popover"]').popover();
    }

    setData(data) {
        this.setState({data: data});
        this.data = data;
    }

}

export default CastList;
