// webpack-php-mock/example/app/components/common/infolist/InfoList.jsx
import React from 'react';
import './InfoList.less';
import loadingGIf from 'imgs/loading-3.gif';

class InfoList extends React.Component {
    nodes = [];

    constructor(props) {
        super(props);
    }

    componentDidMount() {
        var me = this;
        this.$el = $(this.refs.listInfo.getDOMNode());
        this.$nodes = this.$el.find('[data-node$=Panel]');

        this.$nodes.each(function (idx, item) {
            var name = $(item).data('node');
            me.nodes.push(name);
        });
    }

    render() {
        var List = this.props.list;
        var Info = this.props.info;

        return (
            <div className="list-info" ref="listInfo">
                <div className="info swap-item" data-node="infoPanel">
                    <Info {...this.props} ref="info"></Info>
                </div>
                <div className="list swap-item" data-node="listPanel">
                    <List {...this.props} ref="list"></List>
                </div>
                <div className="loading-zoo">
                    <img src={loadingGIf} alt="……加载中……"/>
                </div>
                <div className="detail-btn btn btn-default">
                    <i data-node="detail" className="to-detail glyphicon glyphicon-menu-right"></i>
                </div>
            </div>
        );
    }

    switchTo(name) {
        var idx = this.nodes.indexOf(name + 'Panel');
        idx = -idx;

        this.$nodes.each(function (i, item) {
            $(item).css('transform', 'translateX(' + idx * 100 + '%)')
        });
    }

    inloading() {
        this.$el.addClass('loading');
    }

    unloading() {
        this.$el.removeClass('loading');
    }

    setListData(data) {
        this.refs.list.setData(data);
    }

    setInfoData(data) {
        this.refs.info.setData(data);
    }
}

export default InfoList;
