// webpack-php-mock/example/app/components/common/aoinav/aoinav.jsx
import React from 'react';
import './AoiNav.less';

class AoiNav extends React.Component {
    constructor(props) {
        super(props);
        this.state = {};
    }

    render() {
        return (
            <div className="aoi-nav navbar navbar-default" onChange={this.props.onAoiChange}>
                <a href="javascript:;" className="aoi-brand navbar-brand glyphicon glyphicon-map-marker" data-node="link" data-to="info">
                    北京
                </a>
                <form className="right-gap navbar-form navbar-right">
                    <div className="form-group">
                        <select className="aoi-areas form-control input-sm" 
                            dangerouslySetInnerHTML={{__html: this.state.areas}}
                        >
                        </select>
                    </div>
                </form>
            </div>
        );
    }

    componentDidMount () {
        let me = this;

        if (me.aoiData) {
            me.setAoisState(me.aoiData);
            return;
        }

        $.ajax({
            url: '/logistics/getallaoi',
            dataType: 'json',
            beforeSend: function() {},
            success: function(data) {
                if (0 === data.errno) {
                    me.aoiData = data.data;
                    
                    me.setAoisState(me.aoiData);
                }
            },
            complete: function () {}
        });
    }

    setAoisState(aoiData) {
        var state = {};

        state.areas = `<option value="">全市</option>`;
        state.areas += this.getAreas(aoiData)
            .map(function(item) {
                return `<option value="${item.aoiid}">${item.aoiname}</option>`;
            }).join('');;

        this.setState(state);
    }

    getAreas (data) {
        var oKeys = Object.keys(data);
        var aois = [];

        oKeys.forEach(function (item) {
            aois = data[item].aoi;
        });

        return aois;
    }
}

export default AoiNav;
