// webpack-php-mock/example/app/components/common/cityareas/cityareas.jsx
import React from 'react';

class CityAreas extends React.Component {
    constructor (props) {
        super(props);
        this.state = {};
    }
    getCityies (data) {
        var oKeys = Object.keys(data);

        return oKeys.map(function (item) {
            var city = data[item];
            return {cityname: city.cityname, cityid: city.cityid};
        });
    }
    getAreas (data) {
        var oKeys = Object.keys(data);
        var aois = [];

        oKeys.forEach(function (item) {
            aois = data[item].aoi;
        });

        return aois;
    }
    componentDidMount () {
        let me = this;
        var state = {};

        $.ajax({
            url: '/logistics/getallaoi',
            dataType: 'json',
            beforeSend: function() {},
            success: function(data) {
                if (0 === data.errno) {
                    state.cities = me.getCityies(data.data)
                        .map(function(item) {
                            return `<option value="${item.cityid}">${item.cityname}</option>`;
                        }).join('');
                    state.areas = `<option value="">请选择</option>`;
                    state.areas += me.getAreas(data.data)
                        .map(function(item) {
                            return `<option value="${item.aoiid}">${item.aoiname}</option>`;
                        }).join('');;


                    me.setState(state);
                }
            },
            complete: function () {}
        });

    }
    render () {
        var ss = [];
        if (this.props.city) {
            ss.push(
                <select key="citys" ref="citys" onChange={this.props.onCityChange} className="form-control area-width input-sm" dangerouslySetInnerHTML={{__html: this.state.cities}}>
                </select>
            );
        }

        if (this.props.areas) {
            ss.push(
                <select key="areas" ref="areas" bsSize="small" onChange={this.props.onAreaChange} className="form-control area-width input-sm" dangerouslySetInnerHTML={{__html: this.state.areas}}>
                </select>
            );
        }
        return (
            <span>
                {ss}
            </span>
        );
    }

}

export default CityAreas;
