// webpack-php-mock/example/app/widgets/shopareas/tools/selectbox.jsx
import selectboxEJS from './selectbox.ejs';
import './selectbox.less';

function SelectBox() {
    this.defaultAnchor = BMAP_ANCHOR_TOP_LEFT;
    this.defaultOffset = new BMap.Size(10, 10);
}

SelectBox.prototype = new BMap.Control();

$.extend(SelectBox.prototype, {
    initialize: function(map) {
        var ejs = selectboxEJS();
        var frg = document.createElement('div');
        frg.innerHTML = ejs;
        this.$el = $(map.getContainer());
        this.$el.append(frg);
        this.$aoisEl = this.$el.find('[data-node="aois"]');
        this.$brandsEl = this.$el.find('[data-node="brands"]');
        this.$shopsEl = this.$el.find('[data-node="shops"]');
        this.init();

        return frg;
    },
    init: function() {
        this.setAois();
        this.setBrands();
        this.initEvents();
    },
    initEvents: function () {
        var me = this;
        me.$aoisEl.off().on('change', function (e) {
            var aoiid = e.target.value;
            me.setBrands(aoiid);
        });
    },
    setAois: function() {
        var me = this;
        return $.ajax({
            url: '/logistics/getallaoi',
            dataType: 'json',
            beforeSend: function() {},
            success: function(data) {
                if (0 === data.errno) {
                    var keys = Object.keys(data.data);
                    var aois = [{aoiid: '', aoiname: '请选择'}];
                    var htm = '';
                    for (var i = 0, l = keys.length; i < l; i++) {
                        aois = aois.concat(data.data[keys[i]].aoi);
                    }

                    for (i = 0, l = aois.length; i < l; i++) {
                        htm += `<option value="${aois[i].aoiid}">${aois[i].aoiname}</option>`;
                    }
                    
                    me.$aoisEl.html(htm);
                }
            },
            complete: function () {}
        })
    },
    getAoiid: function() {
        return this.$aoisEl.val();
    },
    setBrands: function(aoiid) {
        var me = this;
        if (!aoiid) {
            return;
        }
        return $.ajax({
            url: '/wlmine/shopoverlapbrand',
            data: {
                aoiid: aoiid
            },
            dataType: 'json',
            beforeSend: function() {},
            success: function(data) {
                if (0 === data.errno) {
                    var bs = ['请选择'];
                    var htm = '';
                    bs = bs.concat(data.data);

                    for (var i = 0, l = bs.length; i < l; i++) {
                        htm += `<option value="${bs[i]}">${bs[i]}</option>`
                    }

                    me.$brandsEl.html(htm);
                }
            },
            complete: function () {}
        })
    },
    brandChange: function(cb) {
        this.$brandsEl.off().on('change', cb)
    },
    getEl: function() {
        return this.$el;
    },
    setShopsBtns: function(shops) {
        var btns = '';
        shops.forEach(function (item, idx) {
            btns += `<button data-sid="${item.wid}" data-node="shop"
                style="background:hsla(${item.idx * 360 / item.total}, 60%, 70%, 0.8)"
            >${item.name}</button>`;
        });
        this.$shopsEl.html(btns);
    }
});

export default SelectBox;
