// webpack-php-mock/example/app/widgets/shopareas/shopareas.jsx
import React from 'react';
import connectToStores from 'alt/utils/connectToStores';
import InfoList from 'components/common/infolist/InfoList';
import AreaList from 'components/shopareas/list/List';
import AreaInfo from 'components/shopareas/info/Info';
import Areas from 'components/shopareas/areas/Areas';
import AreasAction from 'actions/shopareas/AreasAction';
import AreasStore from 'stores/shopareas/AreasStore';
import AoiNav from 'components/common/aoinav/AoiNav';
import ButtonNav from 'components/common/buttonnav/ButtonNav';
import './shopareas.less';
import 'js/map/CurveLine';

@connectToStores
class ShopAreas extends React.Component {
    constructor(props) {
        super(props);
        this.state = {};
    }

    render() {
        return (
            <div id="shop-area" className="info-map" ref="shopArea" onClick={this.clickListInfo}>
                <div className="info-zone">
                    <AoiNav onAoiChange={this.onAoiChange}></AoiNav>
                    <ButtonNav tab={this.state.tab} nav={this.state.nav}></ButtonNav>
                    <InfoList ref="listInfo"
                        list={AreaList} info={AreaInfo}
                        detail={this.state.detail}
                    >
                    </InfoList>
                </div>
                <div className="map-zone">
                    <Areas ref="areasCast"></Areas>
                </div>
            </div>
        );
    }

    componentDidMount() {
        AreasAction.fetchInfo({}, this);
    }

    componentWillReceiveProps(nextProps) {
        var type = nextProps.type;
        if ('list' === type || 'info' === type) {
            this.setState({tab: nextProps.type});
        }
        if ('map' === type) {
            this.setState({castData: nextProps.castData});
        }
    }

    componentWillUpdate(nextProps, nextState) {
        var type = nextProps.type;
        var data;
        var mode;
        var aoiid;

        switch(type) {
            case 'list':
                data = nextProps.listData;
                mode = nextProps.mode;
                aoiid = nextProps.aoiid;
                this.refs.listInfo.setListData({data: data, mode: mode, aoiid: aoiid});
                break;
            case 'info':
                data = nextProps.infoData;
                this.refs.listInfo.setInfoData(data);
            default:
                break;
        }
    }

    componentDidUpdate(prevProps, prevState) {
        var type = this.props.type || 'info';

        switch(type) {
            case 'list':
            case 'info':
                this.refs.listInfo.switchTo(this.props.type);
                break;
            case 'map':
                if ('large' === this.props.mode) {
                    this.refs.areasCast.renderLargeMap(this.props.areasData);
                }
                else {
                    this.refs.areasCast.renderMap(this.props.areasData);
                }
                break;
            default:
                break;
        }
    }

    static getStores(props) {
        return [AreasStore];
    }

    static getPropsFromStores(props) {
        return AreasStore.getState();
    }

    clickListInfo = (evt) => {
        var node = evt.target.dataset.node;

        switch (node) {
            case 'link':
                var to = evt.target.dataset.to;
                var url = evt.target.dataset.url || this.url;
                var mode = evt.target.dataset.mode || this.mode;
                this.url = url || '/wlmine/shopdeliverlist';
                this.mode = mode || 'large';
                if ('list' === to) {
                    AreasAction.fetchList({url: url, aoiid: this.aoiid, mode: mode}, this);
                }
                else {
                    AreasAction.fetchInfo({aoiid: this.aoiid}, this);
                }
                break;
            case 'map':
                var shopId = evt.target.dataset.shopid;
                var brand = evt.target.dataset.brand;
                var mode = evt.target.dataset.mode;
                var aoiid = evt.target.dataset.aoiid;
                AreasAction.fetchAreas({shopid: shopId, mode: mode, brand: brand, aoiid: aoiid}, this);
                break;
            case 'detail':
                this.toggleDetail();
                break;
            default:
                this.to = null;
                break;
        }
    }

    onAoiChange = (evt) => {
        var aoiid = evt.target.value;
        this.aoiid = aoiid;
        if (aoiid) {
            this.setState({nav: true});
        }
        else {
            this.setState({nav: false});
        }
        AreasAction.fetchInfo({aoiid: aoiid}, this);
    }

    toggleDetail() {
        if (this.state.detail) {
            this.closeDetail();
        }
        else {
            this.showDetail();
        }
    }

    showDetail() {
        $(this.refs.shopArea.getDOMNode()).removeClass('info-map').addClass('info-only');
        this.setState({detail: true});
    }

    closeDetail() {
        $(this.refs.shopArea.getDOMNode()).addClass('info-map').removeClass('info-only');
        this.setState({detail: false});
    }
}

export default ShopAreas;
