// webpack-php-mock/example/app/actions/shopcost/CostAction.jsx
/**
 * @file 商户成本数据集中分发
 * @author chenkang01
 */

import flux from 'dispatcher/control';
import {createActions} from 'alt/utils/decorators';
import moment from 'moment';

@createActions(flux)
class CostAction{
    fetchInfo(params, ctx) {
        var me = this;

        $.ajax({
            url: '/wlmine/shopcostoverview',
            data: params,
            dataType: 'json',
            beforeSend: function () {
                ctx.refs.listInfo.inloading();
            },
            success: function (data) {
                if (0 === data.errno) {
                    data.data.isAoi = params.aoiid;
                    me.dispatch(data.data);
                }
            },
            complete: function() {
                ctx.refs.listInfo.unloading();
            }
        });
    }

    fetchList(params, ctx) {
        var me = this;
        $.ajax({
            url: '/wlmine/shopvalueranklistv2',
            data: params,
            dataType: 'json',
            beforeSend: function () {
                ctx.refs.listInfo.inloading();
            },
            success: function (data) {
                if (0 === data.errno) {
                    data = data.data || {};

                    for (var i in data) {
                        if (data.hasOwnProperty(i)) {
                            var desc = data[i].description;
                            var ds = desc.split(';');
                            var a = ds[0];
                            var b = ds[1];
                            var c = data[i].extra_info.direct_advice;

                            var transDesc = [
                                {title: '总体概况', value: [a]},
                                {title: '详细描述', value: [b]},
                                {title: '优化建议', value: [c]}
                            ];
                            data[i].transDesc = transDesc;
                        }
                    }
                    me.dispatch(data);
                }
            },
            complete: function() {
                ctx.refs.listInfo.unloading();
            }
        });
    }

    fetchCast(params, ctx) {
        var me = this;

        params.from = moment(new Date(new Date().setDate((new Date().getDate() - 30)))).format('YYYYMMDD');
        params.to = moment(new Date()).format('YYYYMMDD');

        $.ajax({
            url: '/wlmine/shoporderdistribution',
            data: params,
            dataType: 'json',
            beforeSend: function () {
                ctx.refs.listInfo.inloading();
            },
            success: function (data) {
                if (0 === data.errno) {
                    me.dispatch({data: data.data, shopId: params.shopid});
                }
            },
            complete: function () {
                ctx.refs.listInfo.unloading();
            }
        });
    }
}

export default CostAction;
