// webpack-php-mock/example/app/widgets/layout.jsx
import React from 'react';
import {RouteHandler} from 'react-router';
import Header from 'widgets/header/header';
import Footer from 'widgets/footer/footer';
import Main from 'widgets/main/main';
import Loading from 'components/common/loading/loading';
import toastr from 'toastr';
import 'toastr/toastr.less';
import './layout.less'

class Layout extends React.Component {
    constructor(props) {
        super(props);
        this.setupAjax();
        this.state = {};
    }

    render() {
        return (
          <div id="layout">
            <Header {...this.props}></Header>
            <Main>
                <RouteHandler/>
            </Main>
            <Footer></Footer>
            <Loading show={this.state.loadingShow}></Loading>
          </div>
        );
    }
  
    loading = () => {
        this.setState({loadingShow: true});
    }
  
    unloading = () => {
        this.setState({loadingShow: false});
    }

    setupAjax() {
        var me = this;
        $.ajaxSetup({
            beforeSend: function () {
                me.loading();
            },
            complete: function (jqXHR, status) {
                var resp = jqXHR.responseJSON;
                if (resp && resp.errno !== 0) {
                    toastr.error(resp.errmsg || '系统错误，请稍候重试！');
                }
                me.unloading();
            },
            error: function () {
                toastr.error('系统错误，请稍候重试！');
            }
        });
    }
}

export default Layout;
