// webpack-php-mock/example/app/components/shopareas/areas/shopcubes/ShopCubes.js
import './ShopCubes.less';

function ShopCubes() {
    this.defaultAnchor = BMAP_ANCHOR_TOP_LEFT;
    this.defaultOffset = new BMap.Size(70, 10);
}

ShopCubes.prototype = new BMap.Control();

$.extend(ShopCubes.prototype, {
    initialize: function(map) {
        this.$frg = $(document.createElement('div'));
        this.$frg.addClass('shop-cubes');
        this.$el = $(map.getContainer());
        this.$el.append(this.$frg);
        this.$cubes = this.$el.find('[data-node="cubes"]');
        this.initEvents();

        return this.$frg.get(0);
    },

    clear: function() {
        this.$frg.empty();
    },

    setShops: function(shops) {
        var cubes = shops.map(function (shop) {
            return `<div class="shop" style="background:${shop.color}" data-sid="${shop.wid}">${shop.name}</div>`;
        });
        this.$frg.html(cubes.join(''));
    },

    initEvents: function () {
        var me = this;
        me.$frg.off();
        me.$frg.on('click', '[data-sid]', function (e) {
            var sid = $(e.target).data('sid');
            me.hoverHandler && me.hoverHandler(sid);
        });
    },

    clickShop: function(cb) {
        this.hoverHandler = cb;
    }
});

export default ShopCubes;
