// webpack-php-mock/example/app/actions/shopareas/AreasAction.jsx
/**
 * @file 商户配送范围数据集中分发
 * @author chenkang01
 */

import flux from 'dispatcher/control';
import {createActions} from 'alt/utils/decorators';

@createActions(flux)
class AreasAction{
    fetchInfo(params, ctx) {
        var me = this;

        if (this.ajax) {
            this.ajax.abort();
        }

        this.ajax = $.ajax({
            url: '/wlmine/shopdeliveroverview',
            data: params,
            dataType: 'json',
            beforeSend: function () {
                ctx.refs.listInfo && ctx.refs.listInfo.inloading();
            },
            success: function (data) {
                if (0 === data.errno) {
                    data.data.isAoi = params.aoiid;
                    me.dispatch(data.data);
                }
            },
            complete: function() {
                ctx.refs.listInfo && ctx.refs.listInfo.unloading();
            }
        });
    }

    fetchList(params, ctx) {
        var me = this;

        var url = params.url;
        var mode = params.mode;
        delete params.url;
        delete params.mode;

        if (!url) {
            return;
        }

        $.ajax({
            url: url,
            data: params,
            dataType: 'json',
            beforeSend: function () {
                ctx.refs.listInfo.inloading();
            },
            success: function (data) {
                if (0 === data.errno) {
                    me.dispatch({data: data.data, mode: mode, aoiid: params.aoiid});
                }
            },
            complete: function() {
                ctx.refs.listInfo.unloading();
            }
        });
    }

    fetchAreas(params, ctx) {
        var me = this;

        var mode = params.mode;
        delete params.mode;

        var url;

        if ('overlap' === mode) {
            url = '/wlmine/shopoverlap';
            delete params.shopid;
        }
        else {
            url = '/wlmine/shopdeliverrange';
            delete params.brand;
            delete params.aoiid;
        }

        $.ajax({
            url: url,
            data: params,
            dataType: 'json',
            success: function(data) {
                if (0 === data.errno) {
                    me.dispatch(data.data);
                }
            }
        });
    }
}

export default AreasAction;
